core/destroy.go

```diff
@@ -73,5 +273,149 @@ func (c *CLab) Destroy(ctx context.Context, maxWorkers uint, keepMgmtNet bool) e
 			}
 		}
 	}
+
+	return nil
+}
+
+func (c *CLab) deleteNodes(ctx context.Context, workers uint, serialNodes map[string]struct{}) {
+	wg := new(sync.WaitGroup)
+
+	concurrentChan := make(chan nodes.Node)
+	serialChan := make(chan nodes.Node)
+
+	workerFunc := func(i uint, input chan nodes.Node, wg *sync.WaitGroup) {
+		defer wg.Done()
+		for {
+			select {
+			case n := <-input:
+				if n == nil {
+					log.Debugf("Worker %d terminating...", i)
+					return
+				}
+				err := n.Delete(ctx)
+				if err != nil {
+					log.Errorf("could not remove container %q: %v", n.Config().LongName, err)
+				}
+			case <-ctx.Done():
+				return
+			}
+		}
+	}
+
+	// start concurrent workers
+	wg.Add(int(workers))
+
+	for i := range workers {
+		go workerFunc(i, concurrentChan, wg)
+	}
+
+	// start the serial worker
+	if len(serialNodes) > 0 {
+		wg.Add(1)
+		go workerFunc(workers, serialChan, wg)
+	}
+
+	// send nodes to workers
+	for _, n := range c.Nodes {
+		if _, ok := serialNodes[n.Config().LongName]; ok {
+			serialChan <- n
+			continue
+		}
+		concurrentChan <- n
+	}
+
+	// close channel to terminate the workers
+	close(concurrentChan)
+	close(serialChan)
+
+	// also call delete on the special nodes
+	for _, n := range c.getSpecialLinkNodes() {
+		err := n.Delete(ctx)
+		if err != nil {
+			log.Warn(err)
+		}
+	}
+
+	wg.Wait()
+}
+
+func (c *CLab) deleteToolContainers(ctx context.Context) {
+	toolTypes := []string{"sshx", "gotty"}
+
+	for _, toolType := range toolTypes {
+		toolFilter := []*types.GenericFilter{
+			{
+				FilterType: "label",
+				Field:      containerlablabels.ToolType,
+				Operator:   "=",
+				Match:      toolType,
+			},
+			{
+				FilterType: "label",
+				Field:      containerlablabels.Containerlab,
+				Operator:   "=",
+				Match:      c.Config.Name,
+			},
+		}
+
+		containers, err := c.globalRuntime().ListContainers(ctx, toolFilter)
```

Comment:
I wonder why this is doing list against the runtime rather than using clab list? probably also should reswizzle the label stuff here to use the new option stuff

Reply:
good question that sent me down some memory lane, and I _think_ it was because we might have had in clab some nodes that are not backed by actual containers. This was maybe the case for the multi-node KinD clusters where a single KinD container had several nodes in clab representation (control and worker nodes).

Buuuut, I am so not sure about this :D maybe it is worth checking this up, since we have a kind test in the suites, so if anything goes sideways we hopefully catch it

Reply:
ah, wait wait

in this case the tools containers are launched ad hoc, and not defined in the topology. So when clab runs, it reads the topo file and doesn't find the tools containers. Hence we tagged these containers with labels and runtime returns them based on labels

This goes back to your idea for NewClab to actually be smart and read its state when initializing